src/graph.rs

```diff
@@ -141,44 +160,49 @@ where
             self.reorder(delta_f, delta_b);
         }
 
-        true
+        Ok(())
     }
 
     /// Forwards depth-first-search
     fn dfs_f<'a>(
         &'a self,
-        n: &'a Node<V>,
+        v: V,
         ub: Order,
         visited: &mut HashSet<V>,
-        delta_f: &mut Vec<&'a Node<V>>,
+        delta_f: &mut Vec<&'a Node<V, E>>,
+        cycle: &mut Vec<(V, V, E)>,
     ) -> bool {
+        let n = &self.nodes[&v];
         delta_f.push(n);
 
-        n.out_edges.iter().all(|w| {
-            let node = &self.nodes[w];
-            let ord = node.ord.get();
+        for (w, e) in &n.out_edges {
+            let ord = self.nodes[w].ord.get();
 
             if ord == ub {
                 // Found a cycle
-                false
-            } else if !visited.contains(w) && ord < ub {
+                cycle.push((v, *w, e.clone()));
+                return false;
+            } else if !visited.contains(&w) && ord < ub {
```

Comment:
This is not your doing, but if we're changing the line anyway, we could optimize it to

```suggestion
            } else if ord < ub && visited.insert(w) {
```

and save a hash operation on insertion